### PR TITLE
Try to fix bce loss check

### LINKED ISSUE #41361: Keras model.compile(..., metrics=["accuracy"]) no longer introspects loss function in TF 2.2
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS, Colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): pip/binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b, 2.2.0
- Python version: 3.7.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**

The documentation of [`tf.keras.Model.compile`](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile) includes the following for the `metrics` parameter:

> When you pass the strings 'accuracy' or 'acc', we convert this to one of tf.keras.metrics.BinaryAccuracy, tf.keras.metrics.CategoricalAccuracy, tf.keras.metrics.SparseCategoricalAccuracy based on the loss function used and the model output shape. We do a similar conversion for the strings 'crossentropy' and 'ce' as well. 

The code in question only looks at the model shape, and ignores the loss function:

https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/engine/compile_utils.py#L447-L453

This means that using `acc`/`accuracy` for a model doing binary classification with last dimension > 1 (e.g. binary predictions about multiple independent values for each batch element) will incorrectly use categorical accuracy, not binary accuracy. (That is, the `compile` invocation is equivalent to passing `metrics=["categorical_accuracy"]`.)

**Describe the expected behavior**

The behaviour stated by the documentation is to look at the loss function in addition to the output shape. That is, if the loss function is binary cross-entropy, `metrics=["accuracy"]` should be equivalent to `metrics=["binary_accuracy"]`. 

This is behaviour of TF < 2.2, such as TF 2.1.1:

https://github.com/tensorflow/tensorflow/blob/3ffdb91f122f556a74a6e1efd2469bfe1063cb5c/tensorflow/python/keras/engine/training_utils.py#L1114-L1121

**Standalone code to reproduce the issue**

```python
#%pip install tensorflow==2.2.0
#%pip install tensorflow==2.1.1
import tensorflow as tf

inp = tf.keras.Input(3)
model = tf.keras.Model(inp, inp)

model.compile(loss="bce", metrics=["acc"])
model.evaluate(tf.constant([[0.1, 0.6, 0.9]]), tf.constant([[0, 1, 1]]))

if tf.version.VERSION == "2.2.0":
    print(model.compiled_metrics.metrics[0]._fn.__name__)
else:
    print(model._per_output_metrics[0]["acc"]._fn.__name__)
```

Output

- TF 2.2.0: `categorical_accuracy`
- TF 2.1.1: `binary_accuracy`

Notebook: https://gist.github.com/huonw/4a95b73e3d8a1c48a8b5fc5297d30772
Colab: https://colab.research.google.com/gist/huonw/4a95b73e3d8a1c48a8b5fc5297d30772

**Other info / logs** N/A

### PR DESCRIPTION
**Pull request by bhack**
Fix #41361

### CODE REVIEW

File: tensorflow/python/keras/engine/compile_utils.py [hunk 1]
```diff
@@ -440,8 +441,12 @@ def _get_metric_object(self, metric, y_t, y_p):
       y_p_rank = len(y_p.shape.as_list())
       y_t_last_dim = y_t.shape.as_list()[-1]
       y_p_last_dim = y_p.shape.as_list()[-1]
-
-      is_binary = y_p_last_dim == 1
+      is_binary_crossentropy = (
```
Comment (bhack) [inline tensorflow/python/keras/engine/compile_utils.py:444]: I am not sure that the full set of checks is required

**Comment by bhack** [comment]
Only Windows Bazel failed but it is not related to this PR.

File: tensorflow/python/keras/engine/compile_utils_test.py [hunk 2]
```diff
@@ -531,6 +531,13 @@ def test_accuracy(self):
     self.assertEqual(metric_container.metrics[0]._fn,
                      metrics_mod.binary_accuracy)
 
+    loss = losses_mod.BinaryCrossentropy()
+    metric_container = compile_utils.MetricsContainer('accuracy', loss=loss)
```
Comment (bhack) [inline tensorflow/python/keras/engine/compile_utils_test.py:535]: @huonw Let me know if your case is covered by this test.

Comment (huonw) [reply tensorflow/python/keras/engine/compile_utils_test.py:535]: Looks like it, but there is a reduced test case in #41361 that I cut down from our real code and so is the "true" measure of success. Does that test case match the behaviour of 2.1 with this PR?

Comment (bhack) [reply tensorflow/python/keras/engine/compile_utils_test.py:535]: I've tried that code locally and It Is working with this patch. Can you verify it on your side or check if you want to extend this test coverage.

Comment (huonw) [reply tensorflow/python/keras/engine/compile_utils_test.py:535]: Great, thanks! I'd be happy to say that this PR resolves the issue from our side then.

Unfortunately, I am not in a position to build TensorFlow from source, so unless there's an easy way for me to install a package from CI, I can't confirm on our real code myself. However, this issue was flagged in public code: in particular, https://github.com/stellargraph/stellargraph/issues/1766 found Jupyter notebook where the model reported accuracy metrics of 0 with TF 2.2 (but not TF 2.1). This can be run locally using something like:

1. Download the notebook from https://stellargraph.readthedocs.io/en/v1.2.1/demos/link-prediction/gcn-link-prediction.ipynb
2. Install necessary libraries `pip install stellargraph[demos]==1.2.1`
3. Run the notebook and see whether the `fit` call reports an accuracy of constant zero (bad) or decreasing non-zero (good)

This is a bit fiddly, so I'm not expecting you to do so. 😄

Comment (bhack) [reply tensorflow/python/keras/engine/compile_utils_test.py:535]: >Unfortunately, I am not in a position to build TensorFlow from source, so unless there's an easy way for me to install a package from CI, 

I know that generally It Is resource intensive task and It could be a barrier (E.g. See https://github.com/tensorflow/build/issues/5).

If you see the diff in this PR  is very small so with a little "hack" you can just edit/copy these changes over you installed Tensorflow package. 
You can found where the installed files are with `pip show <your_installed_tensorflow_package_name>`.
Let me know.

Comment (huonw) [reply tensorflow/python/keras/engine/compile_utils_test.py:535]: Ah, good trick. I applied it on Colab: https://colab.research.google.com/gist/huonw/4ac8796f598bc742b476fd8e36a1f866/gcn-link-prediction.ipynb . It didn't work.

Discussion (numbers `[n]` refer to the cell execution count on [the rendered form](https://nbviewer.jupyter.org/gist/huonw/4ac8796f598bc742b476fd8e36a1f866)):

1. I applied the patch with `curl ... | patch` [2]
2. I confirmed the patch was applied with `grep` [3]
3. I ran the example from #41361 with many ways to specify the loss [4]. For example, `"bce"` and referring to the `keras.losses.binary_crossentropy` function directly (the example was extended to include printing the `loss` value used). Output here for convenience:
   ```
   'bce' binary_accuracy
   'binary_crossentropy' categorical_accuracy
   <function binary_crossentropy at 0x7ff1b8db0950> categorical_accuracy
   <tensorflow.python.keras.losses.BinaryCrossentropy object at 0x7ff1a8dc1208> binary_accuracy
   <function binary_crossentropy at 0x7ff1b8db0950> categorical_accuracy
   ```
   Note that only `loss="bce"` and `loss=tf.keras.losses.BinaryCrossentropy()` work, `loss="binary_crossentropy"` and the two forms of `loss=tf.keras.losses.binary_crossentropy` do not.
4. I ran rest of the notebook from https://github.com/stellargraph/stellargraph/issues/1766 without changes to build and compile a Keras model [5]-[19] (using `loss=tf.keras.losses.binary_crossentropy`, as in the original)
5. I inspected the Keras model to confirm that it's using the `MetricsContainer` class with the `loss` parameter to its constructor and the `_loss` attribute [20], [21]
6. I evaluated and trained the model (also without changes from the original notebook), observing that the accuracy was the incorrect 0 value [22]-[25]

File: tensorflow/python/keras/engine/training.py [hunk 3]
```diff
@@ -543,8 +543,11 @@ def compile(self,
       self.optimizer = self._get_optimizer(optimizer)
       self.compiled_loss = compile_utils.LossesContainer(
           loss, loss_weights, output_names=self.output_names)
-      self.compiled_metrics = compile_utils.MetricsContainer(
-          metrics, weighted_metrics, output_names=self.output_names)
+      mc = compile_utils.MetricsContainer(metrics,
+                                          weighted_metrics,
+                                          loss=loss,
```
Comment (huonw) [inline tensorflow/python/keras/engine/training.py:548]: Given my other comment above, maybe this should be using the compiled/normalized `self.compiled_loss` value(s) rather than the generic `loss` one that can be in 'any' form? I don't know what a `LossesContainer` contains in particular, but I'm comparing to the TF 2.1 code, where there didn't need to be a special case for string forms like `"bce"`:

https://github.com/tensorflow/tensorflow/blob/3ffdb91f122f556a74a6e1efd2469bfe1063cb5c/tensorflow/python/keras/engine/training_utils.py#L1114-L1121

Comment (bhack) [reply tensorflow/python/keras/engine/training.py]: OK I modified the patch using the compiled loss. Can you check again?

File: tensorflow/python/keras/engine/compile_utils.py [hunk 4]
```diff
@@ -415,9 +419,22 @@ def update_state(self, y_true, y_pred, sample_weight=None):
   def _get_metric_objects(self, metrics, y_t, y_p):
     """Convert user-supplied metrics to `Metric` objects."""
     metrics = nest.flatten(metrics)
-    return [self._get_metric_object(m, y_t, y_p) for m in metrics]
-
-  def _get_metric_object(self, metric, y_t, y_p):
+    if self._loss_container:
+      losses = self._loss_container._losses
```
Comment (bhack) [inline tensorflow/python/keras/engine/compile_utils.py:423]: @omalleyt12 If all these are required can we extrapolate in a public function in `LossesContainer` outside of `build`. Or is it safe to call the full `build` method directly?

File: tensorflow/python/keras/engine/training.py (hunk 3 above)
Comment (huonw) [reply tensorflow/python/keras/engine/training.py]: I reran the https://colab.research.google.com/gist/huonw/4ac8796f598bc742b476fd8e36a1f866/gcn-link-prediction.ipynb notebook and it seemed to work (other than the change from `model.compiled_metrics._loss` to `model.compiled_metrics._loss_container`). The cell that tests various ways to specify binary crossentropy now has output:

```
'bce' binary_accuracy
'binary_crossentropy' binary_accuracy
<function binary_crossentropy at 0x7f94729b1950> binary_accuracy
<tensorflow.python.keras.losses.BinaryCrossentropy object at 0x7f94629c13c8> binary_accuracy
<function binary_crossentropy at 0x7f94729b1950> binary_accuracy
```

And the accuracy and validation accuracy metrics now look correct too 👍

Comment (bhack) [reply tensorflow/python/keras/engine/training.py]: Thanks for testing

File: tensorflow/python/keras/engine/compile_utils.py (hunk 4 above)
Comment (gbaned) [reply tensorflow/python/keras/engine/compile_utils.py:423]: @omalleyt12 Can you please assist on above comments from @bhack. Thanks!

Comment (gbaned) [reply tensorflow/python/keras/engine/compile_utils.py:423]: @omalleyt12, Any update on this PR? Please. Thanks!

**Comment by bhack** [comment]
Can we close this or are you still interested?

**Comment by bhack** [comment]
@fchollet Just to know cause I spend a little bit of time on this more the one year ago what was the outcome?

**Comment by bhack** [comment]
Also, what about the connect ticket https://github.com/tensorflow/tensorflow/issues/41361?
